x-pack/plugins/security_solution/server/lib/telemetry/endpoint_task.ts

```diff
@@ -136,55 +154,92 @@ export class TelemetryEndpointTask {
       (epMetrics) => {
         return {
           endpoint_agent: epMetrics.latest_metrics.hits.hits[0]._source.agent.id,
+          endpoint_version: epMetrics.latest_metrics.hits.hits[0]._source.agent.version,
           endpoint_metrics: epMetrics.latest_metrics.hits.hits[0]._source,
         };
       }
     );
 
-    if (endpointMetrics.length === 0) {
-      this.logger.debug('no reported endpoint metrics');
-      return 0;
-    }
-
+    /** STAGE 2 - Fetch Fleet Agent Config
+     *
+     * As the policy id + policy version does not exist on the Endpoint Metrics document
+     * we need to fetch information about the Fleet Agent and sync the metrics document
+     * with the Fleet agent's policy data.
+     *
+     * 7.14 ~ An issue was created with the Endpoint agent team to add the policy id +
+     * policy version to the metrics document to circumvent and refactor away from
+     * this expensive join operation.
+     */
     const agentsResponse = endpointData.fleetAgentsResponse;
     if (agentsResponse === undefined) {
-      this.logger.debug('no agents to report');
       return 0;
     }
 
     const fleetAgents = agentsResponse?.agents.reduce((cache, agent) => {
-      cache.set(agent.id, { policy_id: agent.policy_id, policy_version: agent.policy_revision });
+      if (agent.id === '00000000-0000-0000-0000-000000000000') {
+        return cache;
+      }
+
+      if (agent.policy_id !== null && agent.policy_id !== undefined) {
+        cache.set(agent.id, agent.policy_id);
+      }
+
       return cache;
-    }, new Map<string, FleetAgentCacheItem>());
+    }, new Map<string, string>());
 
-    const endpointPolicyCache = new Map<string, FullAgentPolicyInput>();
+    const endpointPolicyCache = new Map<string, PolicyData>();
     for (const policyInfo of fleetAgents.values()) {
-      if (
-        policyInfo.policy_id !== null &&
-        policyInfo.policy_id !== undefined &&
-        !endpointPolicyCache.has(policyInfo.policy_id)
-      ) {
-        const packagePolicies = await this.sender.fetchEndpointPolicyConfigs(policyInfo.policy_id);
-        packagePolicies?.inputs.forEach((input) => {
-          if (input.type === 'endpoint' && policyInfo.policy_id !== undefined) {
-            endpointPolicyCache.set(policyInfo.policy_id, input);
-          }
-        });
+      if (policyInfo !== null && policyInfo !== undefined && !endpointPolicyCache.has(policyInfo)) {
+        const agentPolicy = await this.sender.fetchPolicyConfigs(policyInfo);
+        const packagePolicies = agentPolicy?.package_policies as PackagePolicy[];
+
+        packagePolicies
+          .map((pPolicy) => pPolicy as PolicyData)
```

Comment:
> Is the map just to cast the values, or should this be a filter of any null values? 

It is the former - I am just casting. There shouldn't be any null values (but anything is possible).

> Also, can the pPolicy.inputs array be empty or the config null instead of undefined?

I haven't seen it - as these are package policies instead of agent policies I think they should always have an input. But I'm running with the assumption that it is possible.